src/Features/CSharp/Portable/TypeStyle/UseExplicitTypeCodeFixProvider.cs

```diff
@@ -90,7 +90,8 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
 
             if (parensDesignation is null)
             {
-                var typeSymbol = semanticModel.GetTypeInfo(typeSyntax.StripRefIfNeeded()).ConvertedType;
+                var typeInfo = semanticModel.GetTypeInfo(typeSyntax.StripRefIfNeeded());
+                var typeSymbol = semanticModel.GetNullableContext(node.Span.End).AnnotationsEnabled() ? typeInfo.GetConvertedTypeWithAnnotatedNullability() : typeInfo.ConvertedType;
```

Comment:
The GetConvertedTypeWithAnnotatedNullability should be what ConvertedType is already returning -- is it not in this case? Is this the workaround for the bug you filed?

Reply:
The reason I'm asking is the IDE strips the ?s in unnecessary locations in the simplifier pass here:

https://github.com/dotnet/roslyn/blob/da2ad47c10962b1ac3bc7d093f1f234e1959e815/src/Workspaces/CSharp/Portable/Simplification/Reducers/CSharpNullableAnnotationReducer.cs#L56

So if that's running then we *shouldn't* make this change here. But what I also don't get is GetConvertedTypeWithAnnotatedNullability should be returning what .ConvertedType is also returning after the new symbol wrapping API, so I'd assume this should be a no-op.

Put another way: I can't tell if this is working around a simplifier bug, a compiler bug, both, or neither.

Reply:
@jasonmalinowski Taking the following scenario as an example:
``` csharp
#nullable enable
class Program
{
    void Method(Program x)
    {
#nullable disable
        [|var|] y = x;
        y = null;
    }
}
```
We have the following behavior for `GetTypeInfo` on the `var` type syntax:
- `typeInfo.GetConvertedTypeWithAnnotatedNullability()` returns `Program?`
- `typeInfo.ConvertedType` and `typeInfo.Type` are `Program!`
- `typeInfo.Nullability` has `Annotation: Annotated` and `Flow State: NotNull`

I assumed this was by-design and the purpose of the `GetConvertedTypeWithAnnotatedNullability` method (to stitch the type information with the nullability information provided by `Nullability`). 
Looking at the semantic model implementation (`CSharpSemanticModel.GetTypeInfoForNode`) confirmed my interpretation, we just put the `SourceNamedTypeSymbol` into `TypeInfo.Type` and `TypeInfo.ConvertedType`, which means that it doesn't store any useful top-level nullability information.

Tagging @333fred to advise.

Reply:
Take a look at https://github.com/dotnet/roslyn/pull/40338. The GetConvertedTypeWithAnnotatedNullability() was originally added prior to the compiler having the wrapped APIs; the expectation now is GetTypeInfo() on a type syntax (like var) would return with the annotation that would be applicable automatically.

Reply:
And a bit more history: the only reason GetConvertedTypeWithAnnotatedNullability wasn't already deleted when we moved on to the wrappers was there was some very esoteric uses in extract method we weren't sure about yet. But the intent was to remove if those could be replaced.

Reply:
Digging further, I missed a bit of magic:
```
      public static implicit operator TypeInfo(CSharpTypeInfo info)
        {
            return new TypeInfo(info.Type?.GetITypeSymbol(info.Nullability.FlowState.ToAnnotation()), info.ConvertedType?.GetITypeSymbol(info.ConvertedNullability.FlowState.ToAnnotation()),
                                info.Nullability, info.ConvertedNullability);
        }
```

So the `typeInfo.Type` and `typeInfo.ConvertedType` use the nullability from flow state.
In the `var` scenario the current behavior seems correct. `.Type` and `.ConvertedType` return `Program!` because the flow-state is not-null.

But what I need in this PR is not the type annotated based on flow-state, but rather the type with annotation. So I will need to call `GetConvertedTypeWithAnnotatedNullability()`.

Reply:
@jcouv: so this feature isn't the only feature that would be doing binding on var -- we have (or did...there's a backstory here) other features that call GetTypeInfo() on the type syntax too. So we can't fix in just this one feature I don't think.

It feels very much to me that GetTypeInfo() should be changing behavior here: this language behavior change is in some sense saying "var" means something _other_ than what the flow analysis says. So I'd now expect it to return the forced-null form to match the type of the ILocalSymbol. Put another way, it was OK to use flow state because they were one the in the same. Now that they aren't, shouldn't they be different?

Reply:
And the backstory since that actually might offer some insight: generally a lot of features were always doing GetTypeInfo() on the local's type syntax. In the early Roslyn IDE bringup of Nullable GetTypeInfo() on a type syntax was 100% busted, so I converted some of those to simply grab the local and from there ask for it's type. For example, this bit of this diff:

https://github.com/dotnet/roslyn/pull/36049/files#diff-20970ea4f3b803c82ff01f9921f2014cR2067-R2081

That absolutely wasn't a exhaustive search for everything using that pattern, so I don't know what else is lurking either.

Do we just need to sit down and work through the API impacts here? I feel like trying to figure out the design directly rather than trying to bug stuff into shape in this PR seems wise?

Reply:
> Do we just need to sit down and work through the API impacts here?

I'm tempted to say yes, but I'm afraid of opening a much bigger can of worms than this change warrants. From my understanding, the question of what `GetTypeInfo(...).Type` should return (based on declaration/annotation or null-state?) is not specific to `var`. It is a broader problem. Fixing it is surely a bigger change (if anything updating a large number of tests) and I would not want to do that in 16.5p3.

Would you be okay if we separated that question and tagged this code location with the follow-up issue for `GetTypeInfo`?